library/common/http/dispatcher.cc

```diff
@@ -51,16 +52,60 @@ Dispatcher::DirectStream::DirectStream(envoy_stream_t stream_handle,
     : stream_handle_(stream_handle), underlying_stream_(underlying_stream),
       callbacks_(std::move(callbacks)) {}
 
-Dispatcher::Dispatcher(Event::Dispatcher& event_dispatcher,
-                       Upstream::ClusterManager& cluster_manager)
-    : event_dispatcher_(event_dispatcher), cluster_manager_(cluster_manager) {}
+Dispatcher::Dispatcher() {
+  event_dispatcher_.store(nullptr);
+  cluster_manager_.store(nullptr);
+}
+
+void Dispatcher::ready(Event::Dispatcher& event_dispatcher,
+                       Upstream::ClusterManager& cluster_manager) {
+  {
+    Thread::LockGuard lock(initialization_lock_);
+
+    for (const Event::PostCb& cb : initialization_queue_) {
+      event_dispatcher.post(cb);
+    }
+
+    // The HTTP dispatcher's post function relies on this atomic pointer to signal transition
```

Comment:
FWIW I don't think the complexity of the atomics here is worth it right now. Can you just hold the lock in both init and post and I think that will simplify things substantially?

Reply:
Went ahead with locking; see below.